### DISCUSSION TITLE
How to configure app settings if both using dashboard and api

### DISCUSSION THREAD

**Discussion by gummigroda** [polymind-inc/acmebot#788] [Q&A]
Hi and thanks for a great solution 👍 

I'm trying to configure the app service to allow us to use the dashboard and the api with Entra ID authentication AND only to assigned users/group and applications.  

Reading the appointed documentation: https://github.com/shibayan/keyvault-acmebot/wiki/App-Role-based-authorization
I've added the roles and the App Setting: `Acmebot:AppRoleRequired = true`, I've also added the user I'm using to the roles in the Enterprise App, but trying to issue a certificate in the dashboard gives me a 403. 
Removing the `Acmebot:AppRoleRequired = true` setting and keeping the 'App Role' assignment, lets me again issue certificates.

Regardless, of settings, I'm not able to issue a certificate via below PowerShell snippet:

```pwsh
$at = Get-AzAccessToken -ResourceUrl 'api://6374f649-4f7f-4883-950a-310b24d53fe7' -AsSecureString
$props = @{
  Uri         = "https://REDACTED.azurewebsites.net/api/certificate"
  Headers     = @{Authorization = ('Bearer ' + ($at.Token | ConvertFrom-SecureString -AsPlainText)) }
  Method      = 'POST'
  ContentType = 'application/json; charset=utf-8'
  ErrorAction = 'Stop'
  Body        = @{
    DnsNames = @('test2.domain.tld', 'test3.domain.tld')
  } | ConvertTo-Json -Depth 20
}
Invoke-RestMethod @props -Verbose
```
The error I'm getting when trying the API via PowerShell is: _You do not have permission to view this directory or page._ which seems odd...

How should I configure the App Registration and the Enterprise App to enable access to only specified users to the dashboard and the api for specified applications (SPs)

Thanks!

**Comment by shibayan** [comment]
If you get a 403 error when issuing a certificate from the dashboard, it is likely that the App Role has not been set correctly.

After creating the App Role correctly for the Entra ID App used in App Service Authentication, you need to assign the user application to the Enterprise Application associated with it.

Please check the procedure again.

**Reply by gummigroda** [reply]
__Do I need to create the app roles to be able to access the API?__

As of now, (in a new setup) I have the acmebot _enterprise app_ set to `Assignment required` and added the required users to the `Users and Groups` with the _Default Access_ role. This gives the users access to the Dashboard = All Good.

Now, when I try to access the API via powershell, I get the same error as above: `You do not have permission to view this directory or page`.
Usually, on my own Powershell function apps, this exact setup allows me to access the API as one of the assigned users. Perhaps I'm requesting the access token the incorrect way?

Is the `Acmebot:AppRoleRequired` setting and the setup of the AppRoles, just a way to delegate different permissions or is it __NEEDED__ to be able to access the API and dashboard 'simultaneous'?

So my question is really, if we don't need different appRoles (all users are allowed to issue and revoke), what's needed to be able to access the dashboard and the API as an assigned user?

**Reply by gummigroda** [reply]
A quick update, I've tried to do a new deployment in a test environment via the `Deploy To Azure` button, and then added EasyAuth via the Portal within the Function. Same error as described above.
Then tried the Terraform example from here: https://github.com/shibayan/terraform-azurerm-keyvault-acmebot/blob/master/example/main.tf in same test environment.
Had some initial deployment problems, but after a successful deployment, I'm able to access the API. I'm able to see some differences in the terraform deployed app reg/enterprise app versus the 'Portal enabled one' within the function app. Let's see if I find the culprit. 
Any ideas?

**Comment by igoltz** [comment]
Hello, thanks for the great tool!
I do follow up here instead of creating a new question.
We use it via GUI but now need some automation. But can't get the API access to work. Followed the documentation and other references like https://github.com/SparebankenVest/ingress2acmebotreflector/blob/main/controllers/ingress_controller.go
But all we get is "You do not have permission to view this directory or page." when using the API.
Simple test:

```bash
TENNANTID=ourTennent
CLIENTSECRET=xxxManuallyCreatedServicePrinicpal
APPCLIENTID=xxxManuallyCreatedServicePrinicpal
SCOPE=api://yyyAppRegDeployedByAcmebot/.default

TOKEN=$(curl -s -H 'Content-Type: application/x-www-form-urlencoded' \
-d "client_id=${APPCLIENTID}&scope=${SCOPE}&grant_type=client_credentials&client_secret=${CLIENTSECRET}" \
-X POST "https://login.microsoftonline.com/${TENNANTID}/oauth2/v2.0/token" \
| jq .access_token)

curl -H "Accept: application/json"  -H 'Content-Type: application/json' -H "Authorization: Bearer ${TOKEN}" \
-X GET "https://func-fn-acmeboyyyy.azurewebsites.net/api/certificates"
```
We configured app roles on the functions registration
<img width="1304" height="494" alt="approles" src="https://github.com/user-attachments/assets/13233a76-4162-4304-b63a-a2e761e4f3bb" />

Gave the manually created registration (the one used for client auth.) API permissions
<img width="1380" height="770" alt="spAPIperm" src="https://github.com/user-attachments/assets/fa18325b-4ee9-4d3a-8d52-c720dcb9e490" />
These are seen now in the functions enterprise application
<img width="1627" height="492" alt="fnEAUsers" src="https://github.com/user-attachments/assets/836e46d2-5727-4637-a20b-7ab8ca07af71" />

The functions Auth. settings are:
<img width="573" height="584" alt="fnAuthSettings" src="https://github.com/user-attachments/assets/82e27ecd-5728-4440-b08e-4abc8bba95ff" />

Following the above description also tried to add new paramter:
 
<img width="1072" height="649" alt="envSetting" src="https://github.com/user-attachments/assets/32489e56-55e3-40c9-b542-e3bc784ff8a5" />


What are we missing, doing complete wrong? Your help is much appreciated.
Thanks a lot!
